### ISSUE TITLE
FreeBSD support for LXQt

### ISSUE THREAD

**Issue by BabylonAS** [lxqt/lxqt#1478]
Hello everyone.

I am currently considering to use LXQt for my ancient, early 2000s laptop with FreeBSD 11 in place of the MATE desktop, which appears to be laggy and unwieldy for a roughly 15-16 years old small computer. But it seems like the project is not quite aimed towards BSD-like systems. I have only found four LXQt-related packages in the official FreeBSD package repository, and my attempts to compile the repository manually meet unresolvable problems, apparently because the scripts are expecting to use Linux versions of certain utilities like ``make``.

I would like to make sure the LXQt project is compatible with FreeBSD. Although I hope the problems are exclusively related to the build process, if the problem is deeper, I would like to know. By the way, I have some experience of programming with Qt, so maybe I can do something beneficial to this project.

Thanks in advance!

**Comment by tsujan** [comment]
>  if the problem is deeper, I would like to know...

As far as I know -- and I may be wrong -- none of us has BSD but LXQt should work on BSD. So, it seems that you're the only person who could really answer that question.

> By the way, I have some experience of programming with Qt, so maybe I can do something beneficial to this project.

It would be great if you could find BSD-specific problems and fix them in PRs :)

**Comment by BabylonAS** [comment]
Apparently I could’t build it because of problems with the make’s -j parameter. It specifies the number of jobs, and by default it would be retrieved by GNU coreutils command `nproc`, but it’s not installed. The problem is circumvented by hard-coding the number of jobs, however while on `bash` it’s easy, on `tcsh` (default shell I use on FreeBSD) it’s problematic, so I tried to run the `build_all.sh` script from `bash` with `LXQT_JOB_NUM=1`. Right now it works, I only meet some dependency-related issues, like the K Window Manager not being installed, but expect me reporting some issues not related to dependencies in real time.

Also, the shebang lines should use `#!/usr/bin/env bash` instead of hard-coded `#!/bin/bash`, because FreeBSD puts `bash` at `/usr/local/bin/bash`. Thankfully, `/usr/bin/env` is on the same directory on both systems.

**Comment by tsujan** [comment]
Just a suggestion (ignore it if irrelevant):

If I had BSD, I'd try to install the LXQt components one by one and directly (preferably, by making installable packages), in the order explained here: https://github.com/lxqt/lxqt/wiki/Building-from-source

**Comment by agaida** [comment]
please be a good boy and never ever try build_all.sh on *BSD - it is pain in the arse even with linux
The right way™ would be - read the wiki and esp. the build order and build LXQt in that regard. :D

**Comment by agaida** [comment]
hmm - additional - there was some efforts by several people to make LXQt work on *BSD - should be no problem at all - qterminal will be a little bit funny. At least the combination FreeBSD kernel mixed with a GNU userland works fine, i run this kombi for 2 years now

**Comment by agaida** [comment]
@BabylonAS - overlapping posts - but @tsujan is right, this is the only sane way to do it.

**Comment by BabylonAS** [comment]
@tsujan Thanks for the link. I will consider it.

**Comment by tsujan** [comment]
@BabylonAS Please tell us about the result or whether any change is needed for *BSD.

**Comment by BabylonAS** [comment]
Building `fm-qt` fails because it uses a GNU-supplied `mempcpy(3)` function which is not present on my system. Installing `pcmanfm-qt` (is it the same?) didn't solve the issue. Also, for some reason libkscreen (one of the dependencies) required Qt 4 and some KDE 4 libraries, yet the project is oriented for the fifth version.

**Comment by BabylonAS** [comment]
I've managed to work around the libfm-qt issue by overriding the CMake repo lists. Perhaps the biggest problem right now is related to the `libpulse` library: `pavucontrol-qt` final linkage fails with `-lpulse-mainloop-glib` not found, yet the library is present at `/usr/local/lib/libpulse-mainloop-glib.so.0.0.5` (installed as part of the `pulseaudio` package).

**Comment by tsujan** [comment]
@BabylonAS If you find solutions compatible with Linux, please make PRs.

**Comment by BabylonAS** [comment]
@tsujan I wasn’t talking about Linux. I need LXQt on FreeBSD, and I need to know why `pavucontrol-qt`’s final linking has failed. I don’t exclude that might be related to discrepancies between gcc’s and clang’s parameter evaluation.

**Comment by BabylonAS** [comment]
How? I don’t have any plausible ideas. I’m dealing with CMake for the first time, and I am patching up a complex project downloaded from elsewhere at all for the first time. As my bio on the profile states, I’m an amateur programmer, if not lower, although I may end up to be a good learner.

**Comment by tsujan** [comment]
@BabylonAS That was a simple request, not a demand -- I used the word "please", not "should" ;)

**Comment by BabylonAS** [comment]
Still I don’t have ideas. On the other hand, my question about `-lpulse-mainloop-glib` didn’t get a decisive answer.

**Comment by palinek** [comment]
> I've managed to work around the libfm-qt issue by overriding the CMake repo lists. Perhaps the biggest problem right now is related to the libpulse library: pavucontrol-qt final linkage fails with -lpulse-mainloop-glib not found, yet the library is present at /usr/local/lib/libpulse-mainloop-glib.so.0.0.5 (installed as part of the pulseaudio package).

1. do the final build step with `make VERBOSE=1` to see all the commands invoked
2. the linker doesn't look for the `lib*.so.*` libraries, but for `lib*.so`, which are usually a symlinks to the the "real" (versioned) files. The symlinks are usually parts of the `*-dev` packages

**Comment by BabylonAS** [comment]
I’ve found out the linking command’s code without VERBOSE=1 by thoroughly examining Makefiles and CMake files. Located at pavucontrol-qt/build/src/CMakeFiles/pavucontrol-qt.dir/link.txt, its contents are:
``
/usr/bin/c++   -fno-exceptions -Wall -Wextra -Wchar-subscripts -Wno-long-long -Wpointer-arith -Wundef -Wformat-security -Wnon-virtual-dtor -Woverloaded-virtual -Wpedantic -Wno-gnu-zero-variadic-macro-arguments -std=c++11 -g  -Wl,-Bsymbolic-functions -Wl,-Bsymbolic  CMakeFiles/pavucontrol-qt.dir/pavucontrol.cc.o CMakeFiles/pavucontrol-qt.dir/mainwindow.cc.o CMakeFiles/pavucontrol-qt.dir/cardwidget.cc.o CMakeFiles/pavucontrol-qt.dir/channel.cc.o CMakeFiles/pavucontrol-qt.dir/devicewidget.cc.o CMakeFiles/pavucontrol-qt.dir/minimalstreamwidget.cc.o CMakeFiles/pavucontrol-qt.dir/rolewidget.cc.o CMakeFiles/pavucontrol-qt.dir/sinkinputwidget.cc.o CMakeFiles/pavucontrol-qt.dir/sinkwidget.cc.o CMakeFiles/pavucontrol-qt.dir/sourceoutputwidget.cc.o CMakeFiles/pavucontrol-qt.dir/sourcewidget.cc.o CMakeFiles/pavucontrol-qt.dir/streamwidget.cc.o CMakeFiles/pavucontrol-qt.dir/pavucontrol-qt_autogen/mocs_compilation.cpp.o  -o pavucontrol-qt -Wl,-rpath,/usr/local/lib/qt5:/usr/local/lib: /usr/local/lib/qt5/libQt5Widgets.so.5.9.4 /usr/local/lib/qt5/libQt5DBus.so.5.9.4 -lpulse-mainloop-glib -lpulse -lpthread -lglib-2.0 -lintl -lglib-2.0 -lintl /usr/local/lib/qt5/libQt5Gui.so.5.9.4 /usr/local/lib/qt5/libQt5Core.so.5.9.4 -Wl,-rpath-link,/usr/local/lib
``
(Note that I am using clang++.)

**Comment by palinek** [comment]
It seems that linker on your system doesn't look into `/usr/local/lib` by default and there is no `-L/urs/local/lib` flag in the linker command. @BabylonAS  can you check if the lxqt/pavucontrol-qt#47 makes this better?

**Comment by BabylonAS** [comment]
@palinek The fix did work, actually. Thanks!
@tsujan I wasn’t aware of that page.

**Comment by BabylonAS** [comment]
Now I am running into problems with kf5screen: it is not available in the package repositories, and `libkscreen` is apparently suited for KDE 4.

**Comment by BabylonAS** [comment]
Okay, so the missing component is the `x11/plasma5-libkscreen` package, which is present neither in the packages nor in the ports (yet I have updated the ports collection not long ago). Is there some way of installing it on FreeBSD?

**Comment by jsm222** [comment]
Check out wiki.freebsd.org/LXQt based on work of Olivier Duchateau and me, porting to FreeBSD is happening.

**Comment by BabylonAS** [comment]
Never mind, I decided not to use the lxqtmerge script and have successfully built KF5's libkscreen directly from a repository hosted here, on GitHub.

After this, and a problem with another module similar to the `pavucontrol`'s issue (which, this time, I have fixed myself), I have stumbled on the `lxqt-session` module, which apparently requires UDev (something not available on FreeBSD).

**Comment by jsm222** [comment]
@BabylonAS In the upcoming ports we use CMAKE_ARGS+=    -DWITH_LIBUDEV:BOOL=OFF for lxqt-session. If you want to use lxqtmerge.sh set PORTSDIR (by editing lxqtmerge.sh) to where your clean checked out portstree is. e.g normally the default /usr/ports

**Comment by BabylonAS** [comment]
@jsm222 Thanks. The `lxqtmerge` script didn’t actually work with `PORTSDIR=/usr/ports`, but that doesn’t bother me anymore because, once again, I decided to not use it.

**Comment by BabylonAS** [comment]
`lxqt-sudo` (specifically `sudo.cpp`) requires `<pty.h>`, which is not present on FreeBSD, but everything it needs from there is available in three other headers. I have manually changed the code to include these headers (`<libutil.h>`, `<termios.h>`, and `<errno.h>`), and the program compiled successfully. I have also found out how to ignore the Linux-specific `lm_sensors` module when building `lxqt-panel` (however it failed to build for different reasons).

**Comment by jsm222** [comment]
FYI you are mostly ending up with duplicate work of https://github.com/jsm222/lxqt-freebsd/tree/0.12. So why not just check that out, and report your problems with that? Follow instructions on the FreeBSD Wiki page. Of course it is fully up to you.

**Comment by jsm222** [comment]
Well we are waiting to be imported in the ports tree, which depends on the import of https://github.com/freebsd/freebsd-ports-kde/tree/kde5-import/ 
Meanwhile the wiki.FreeBSD.org/LXQt documents how to install LXQt on FreeBSD. So yes the work is already done.

**Comment by tsujan** [comment]
@jsm222 Thanks! Happy to know that.

**Comment by BabylonAS** [comment]
No, wait, I am encountering yet another problem with `lxqt-panel`. Like `pavucontrol-qt`, it can't find a library (`libXcomposite.so`), but I can't find any CMake directive that controls this.

**Comment by BabylonAS** [comment]
Never mind, I did a brute-force search with `find` and found the problematic `CMakeFiles.txt` file (it was related to `plugin-tray`). Some other modules with similar linkage problems were rather easy to fix.

With the final components compiled, **I have finished building LXQt on FreeBSD!** In comparison to MATE, the new enviroment is really faster and less memory-intensive (about 270 MiB of RAM usage, compared to 400-plus for MATE). Thanks to anyone who helped me in making the environment work!
